File: lab4/FA.py
class FiniteAutomata:

    def __init__(self, Q, E, q0, F, S):
        self.Q = Q
        self.E = E
        self.q0 = q0
        self.F = F
        self.S = S

    @staticmethod
    def getLine(line):
        return line.strip().split(' ')[2:]

    @staticmethod
    def validate(Q, E, q0, F, S):
        if q0 not in Q:
            return False
        for state in F:
            if state not in Q:
                return False
        for key in S.keys():
            if key[0] not in Q:
                return False
            if key[1] not in E:
                return False
            for destination in S[key]:
                if destination not in Q:
                    return False
        return True

    @staticmethod
    def readFromFile(filename):
        with open(filename) as file:
            Q = FiniteAutomata.getLine(file.readline())
            E = FiniteAutomata.getLine(file.readline())
            q0 = FiniteAutomata.getLine(file.readline())[0]
            F = FiniteAutomata.getLine(file.readline())

            file.readline()
            S = {}
            for line in file:
                source = line.strip().split('->')[0].strip().replace('(', '').replace(')', '').split(',')[0]
                route = line.strip().split('->')[0].strip().replace('(', '').replace(')', '').split(',')[1]
                destination = line.strip().split('->')[1].strip()

                if (source, route) in S.keys():
                    S[(source, route)].append(destination)
                else:
                    S[(source, route)] = [destination]

            if not FiniteAutomata.validate(Q, E, q0, F, S):
                raise Exception("Invalid finite automata!")

            return FiniteAutomata(Q, E, q0, F, S)

    def isDfa(self):
        for k in self.S.keys():
            if len(self.S[k]) > 1:
                return False
        return True

    def isAccepted(self, sequence):
        if self.isDfa():
            currentState = self.q0
            for symbol in sequence:
                if (currentState, symbol) in self.S.keys():
                    currentState = self.S[(currentState, symbol)][0]
                else:
                    return False
            return currentState in self.F
        return False

    def __str__(self):
        return "Q = { " + ', '.join(self.Q) + " }\n" \
                "E = { " + ', '.join(self.E) + " }\n" \
                "q0 = { " + self.q0 + " }\n" \
                "F = { " + ', '.join(self.F) + " }\n" \
                "S = " + str(self.S)

File: lab2/src/FA.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.*;

public class FA {
    private final List<String> Q, E, F;
    private String q0;
    private Map<AbstractMap.SimpleEntry<String, String>, List<String>> S;

    public FA() {
        this.Q = new ArrayList<>();
        this.E = new ArrayList<>();
        this.F = new ArrayList<>();
        this.q0 = "";
        this.S = new HashMap<>();
    }

    public FA(List<String> Q, List<String> E, String q0, List<String> F, Map<AbstractMap.SimpleEntry<String, String>, List<String>> S) {
        this.Q = Q;
        this.E = E;
        this.q0 = q0;
        this.F = F;
        this.S = S;
    }

    public FA(String filename) {
        this.Q = new ArrayList<>();
        this.E = new ArrayList<>();
        this.F = new ArrayList<>();
        this.S = new HashMap<>();
        readFromFile(filename);
    }

    public List<String> getQ() {
        return Q;
    }

    public List<String> getE() {
        return E;
    }

    public String getQ0() {
        return q0;
    }

    public List<String> getF() {
        return F;
    }

    public Map<AbstractMap.SimpleEntry<String, String>, List<String>> getS() {
        return S;
    }

    public boolean validate() {
        if (!Q.contains(q0)) {
            return false;
        }
        for (var state : F) {
            if (!Q.contains(state)) {
                return false;
            }
        }
        for (var key: S.keySet()){
            if (!Q.contains(key.getKey())) {
                return false;
            }
            if (!E.contains(key.getValue())) {
                return false;
            }
            for (var destination : S.get(key)) {
                if (!Q.contains(destination)) {
                    return false;
                }
            }
        }
        return true;
    }

    public void readFromFile(String filename) {
        try{
            File file = new File(filename);
            BufferedReader br = new BufferedReader(new FileReader(file));

            String line = br.readLine();
            if (line != null){
                Q.addAll(Arrays.asList(line.split(" ")));
            }
            line = br.readLine();
            if (line != null){
                E.addAll(Arrays.asList(line.split(" ")));
            }
            line = br.readLine();
            if (line != null){
                q0 = line.strip();
            }
            line = br.readLine();
            if (line != null){
                F.addAll(Arrays.asList(line.split(" ")));
            }
            line = br.readLine();
            while (line != null){
                String[] tokens = line.split(" ");
                AbstractMap.SimpleEntry<String, String> key = new AbstractMap.SimpleEntry<>(tokens[0], tokens[1]);
                if(S.containsKey(key)){
                    S.get(key).add(tokens[2]);
                }
                else{
                    ArrayList<String> value = new ArrayList<>(Collections.singletonList(tokens[2]));
                    S.put(key, value);
                }
                line = br.readLine();
            }

        } catch (Exception e){
            e.printStackTrace();
        }
    }

    public boolean isDfa() {
        for (var key : S.keySet()) {
            if (S.get(key).size() > 1) {
                return false;
            }
        }
        return true;
    }

    public boolean isAccepted(String sequence) {
        if(isDfa()){
            var currentState = q0;
            for(char symbol : sequence.toCharArray()){
                var key = new AbstractMap.SimpleEntry<>(currentState, String.valueOf(symbol));
                if(S.containsKey(key)){
                    currentState = S.get(key).get(0);
                }
                else {
                    return false;
                }
            }
            return F.contains(currentState);
        }
        return false;
    }

    @Override
    public String toString() {
        return "Q = { " + String.join(", ", Q) + " }\n" +
                "E = { " + String.join(", ", E) + " }\n" +
                "q0 = { " + q0 + " }\n" +
                "F = { " + String.join(", ", F) + " }\n" +
                "S = " + S;
    }
}
